cmd/cosign/cli/verify/verify_bundle.go

```diff
@@ -149,15 +164,159 @@ func verifyNewBundle(ctx context.Context, bundlePath, trustedRootPath, keyRef, s
+	payload, err := payloadBytes(artifactRef)
+	if err != nil {
+		return nil, err
+	}
+	buf := bytes.NewBuffer(payload)
+	digest := sha256.Sum256(buf.Bytes())
+
+	pb := &protobundle.Bundle{
+		MediaType:            "application/vnd.dev.sigstore.bundle+json;version=0.3",
+		VerificationMaterial: &protobundle.VerificationMaterial{},
+	}
+
+	if envelope != nil && len(envelope.Signatures) > 0 {
+		sigDecode, err := base64.StdEncoding.DecodeString(envelope.Signatures[0].Sig)
+		if err != nil {
+			return nil, err
+		}
+
+		sig := &protodsse.Signature{
+			Sig: sigDecode,
+		}
+
+		payloadDecode, err := base64.StdEncoding.DecodeString(envelope.Payload)
+		if err != nil {
+			return nil, err
+		}
+
+		pb.Content = &protobundle.Bundle_DsseEnvelope{
+			DsseEnvelope: &protodsse.Envelope{
+				Payload:     payloadDecode,
+				PayloadType: envelope.PayloadType,
+				Signatures:  []*protodsse.Signature{sig},
+			},
+		}
+	} else {
+		pb.Content = &protobundle.Bundle_MessageSignature{
+			MessageSignature: &protocommon.MessageSignature{
+				MessageDigest: &protocommon.HashOutput{
+					Algorithm: protocommon.HashAlgorithm_SHA2_256,
+					Digest:    digest[:],
+				},
+				Signature: sigBytes,
+			},
+		}
+	}
+
+	if cert != nil {
+		pb.VerificationMaterial.Content = &protobundle.VerificationMaterial_Certificate{
+			Certificate: &protocommon.X509Certificate{
+				RawBytes: cert.Raw,
+			},
+		}
+	} else if sigVerifier != nil {
+		pub, err := sigVerifier.PublicKey(pkOpts...)
+		if err != nil {
+			return nil, err
+		}
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(pub)
+		if err != nil {
+			return nil, err
+		}
+		hashedBytes := sha256.Sum256(pubKeyBytes)
+
+		pb.VerificationMaterial.Content = &protobundle.VerificationMaterial_PublicKey{
+			PublicKey: &protocommon.PublicKeyIdentifier{
+				Hint: base64.StdEncoding.EncodeToString(hashedBytes[:]),
+			},
+		}
+	}
+
+	if len(signedTimestamp) > 0 {
+		ts := &protocommon.RFC3161SignedTimestamp{
+			SignedTimestamp: signedTimestamp,
+		}
+
+		pb.VerificationMaterial.TimestampVerificationData = &protobundle.TimestampVerificationData{
+			Rfc3161Timestamps: []*protocommon.RFC3161SignedTimestamp{ts},
+		}
+	}
+
+	if !ignoreTlog {
+		var pem []byte
+		var err error
+		if cert != nil {
+			pem, err = cryptoutils.MarshalCertificateToPEM(cert)
+			if err != nil {
+				return nil, err
+			}
+		} else if sigVerifier != nil {
+			pub, err := sigVerifier.PublicKey(pkOpts...)
+			if err != nil {
+				return nil, err
+			}
+			pem, err = cryptoutils.MarshalPublicKeyToPEM(pub)
+			if err != nil {
+				return nil, err
+			}
+		}
+		var sigB64 string
+		var payload []byte
+		if envelope != nil && len(envelope.Signatures) > 0 {
+			payload, err = json.Marshal(*envelope)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			sigB64 = base64.StdEncoding.EncodeToString(sigBytes)
+			payload = buf.Bytes()
+		}
+
+		tlogEntries, err := cosign.FindTlogEntry(ctx, rekorClient, sigB64, payload, pem)
```

Comment:
Something seems to be wrong with how the signature is being encoded when I try to use the detached signature generated by cosign:

```
$ go run cmd/cosign/main.go sign-blob /tmp/blob -y --output-signature /tmp/blob.sig  --key cosign.key 
Using payload from: /tmp/blob
[snipped]
tlog entry created with index: 144824671
Wrote signature to file /tmp/blob.sig
$ go run cmd/cosign/main.go bundle create --signature /tmp/blob.sig --out /tmp/test.converted.bundle --artifact /tmp/blob --key cosign.pub 
Error: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating IEEE_P1363 encoded signature}
main.go:74: error during command execution: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating IEEE_P1363 encoded signature}
exit status 1
$ go run cmd/cosign/main.go sign-blob /tmp/blob -y --output-signature /tmp/blob.sig  --key cosign.key --b64
Using payload from: /tmp/blob
[snipped]
tlog entry created with index: 144825079
Wrote signature to file /tmp/blob.sig
$ go run cmd/cosign/main.go bundle create --signature /tmp/blob.sig --out /tmp/test.converted.bundle --artifact /tmp/blob --key cosign.pub 
Error: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating IEEE_P1363 encoded signature}
main.go:74: error during command execution: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: invalid signature when validating IEEE_P1363 encoded signature}
exit status 1
```

Reply:
If you want to confirm what's been uploaded against what's being constructed, the `--bundle` can contain the canonicalized entry - https://github.com/sigstore/cosign/blob/9b17791a21f9eef8ff59e6ea138e1b116dc571b8/pkg/cosign/bundle/rekor.go#L29

Reply:
Oops! Yes, this is another bug in handling detached signatures, which should now be fixed.